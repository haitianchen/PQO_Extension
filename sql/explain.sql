CREATE EXTENSION sr_plan;

SET sr_plan.log_usage = NOTICE;
CREATE TABLE explain_test(test_attr1 int, test_attr2 int);
INSERT INTO explain_test SELECT i, i + 1 FROM generate_series(1, 20) i;

SET sr_plan.write_mode = true;
SELECT * FROM explain_test WHERE test_attr1 = 10;

SET sr_plan.write_mode = false;
UPDATE sr_plans SET enable = true;

-- check show_plan
WITH vars AS (SELECT query_hash FROM sr_plans WHERE
	query = 'SELECT * FROM explain_test WHERE test_attr1 = 10;' LIMIT 1)
SELECT show_plan(vars.query_hash) FROM vars;

WITH vars AS (SELECT query_hash FROM sr_plans WHERE
	query = 'SELECT * FROM explain_test WHERE test_attr1 = 10;' LIMIT 1)
SELECT show_plan(vars.query_hash, index := 1) FROM vars;

WITH vars AS (SELECT query_hash FROM sr_plans WHERE
	query = 'SELECT * FROM explain_test WHERE test_attr1 = 10;' LIMIT 1)
SELECT show_plan(vars.query_hash, index := 2) FROM vars;

WITH vars AS (SELECT query_hash FROM sr_plans WHERE
	query = 'SELECT * FROM explain_test WHERE test_attr1 = 10;' LIMIT 1)
SELECT show_plan(vars.query_hash, format := 'json') FROM vars;

WITH vars AS (SELECT query_hash FROM sr_plans WHERE
	query = 'SELECT * FROM explain_test WHERE test_attr1 = 10;' LIMIT 1)
SELECT show_plan(vars.query_hash, format := 'text') FROM vars;

WITH vars AS (SELECT query_hash FROM sr_plans WHERE
	query = 'SELECT * FROM explain_test WHERE test_attr1 = 10;' LIMIT 1)
SELECT show_plan(vars.query_hash, format := 'xml') FROM vars;

WITH vars AS (SELECT query_hash FROM sr_plans WHERE
	query = 'SELECT * FROM explain_test WHERE test_attr1 = 10;' LIMIT 1)
SELECT show_plan(vars.query_hash, format := 'yaml') FROM vars;

WITH vars AS (SELECT query_hash FROM sr_plans WHERE
	query = 'SELECT * FROM explain_test WHERE test_attr1 = 10;' LIMIT 1)
SELECT show_plan(vars.query_hash, format := 'nonsense') FROM vars;

DROP TABLE explain_test CASCADE;
DROP EXTENSION sr_plan CASCADE;
