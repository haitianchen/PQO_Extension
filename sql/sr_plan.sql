CREATE EXTENSION sr_plan;

SET sr_plan.log_usage = NOTICE;
CREATE TABLE test_table(test_attr1 int, test_attr2 int);
INSERT INTO test_table SELECT i, i + 1 FROM generate_series(1, 20) i;
SET sr_plan.write_mode = true;
SELECT * FROM test_table WHERE test_attr1 = _p(10);
SELECT * FROM test_table WHERE test_attr1 = 10;
SELECT * FROM test_table WHERE test_attr1 = 10;
SET sr_plan.write_mode = false;

SELECT * FROM test_table WHERE test_attr1 = _p(10);
SELECT * FROM test_table WHERE test_attr1 = 10;
SELECT * FROM test_table WHERE test_attr1 = 15;

UPDATE sr_plans SET enable = true;


SELECT * FROM test_table WHERE test_attr1 = _p(10);
SELECT * FROM test_table WHERE test_attr1 = _p(15);
SELECT * FROM test_table WHERE test_attr1 = 10;
SELECT * FROM test_table WHERE test_attr1 = 15;

SELECT enable, query FROM sr_plans ORDER BY length(query);
DROP TABLE test_table;
SELECT enable, query FROM sr_plans ORDER BY length(query);

CREATE TABLE test_table(test_attr1 int, test_attr2 int, test_attr3 int);

SELECT * FROM test_table WHERE test_attr1 = _p(10);
SELECT * FROM test_table WHERE test_attr1 = 10;
SELECT * FROM test_table WHERE test_attr1 = 10;
SELECT * FROM test_table WHERE test_attr1 = 15;

DROP EXTENSION sr_plan CASCADE;
DROP TABLE test_table;
