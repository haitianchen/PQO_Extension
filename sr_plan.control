# sr_plan extension
comment = 'functions for save and read plan'
default_version = '1.2'
module_pathname = '$libdir/sr_plan'
