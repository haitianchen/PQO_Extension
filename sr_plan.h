#ifndef ___SR_PLAN_H__
#define ___SR_PLAN_H__

#include "postgres.h"
#include "fmgr.h"
#include "string.h"
#include "optimizer/planner.h"
#include "nodes/print.h"
#include "nodes/makefuncs.h"
#include "nodes/nodeFuncs.h"
#include "utils/jsonb.h"
#include "utils/builtins.h"
#include "utils/rel.h"
#include "utils/relcache.h"
#include "utils/tqual.h"
#include "utils/guc.h"
#include "utils/datum.h"
#include "utils/inval.h"
#include "utils/snapmgr.h"
#include "utils/fmgroids.h"
#include "portability/instr_time.h"
#include "storage/lock.h"
#include "access/heapam.h"
#include "access/htup_details.h"
#include "parser/analyze.h"
#include "parser/parse_func.h"
#include "tcop/utility.h"
#include "catalog/pg_type.h"
#include "commands/explain.h"
#include "utils/syscache.h"
#include "funcapi.h"

#define SR_PLANS_TABLE_NAME "sr_plans"
#define SR_PLANS_TABLE_QUERY_INDEX_NAME "sr_plans_query_hash_idx"
#define SR_PLANS_RELOIDS_INDEX "sr_plans_query_oids"
#define SR_PLANS_INDEX_RELOIDS_INDEX "sr_plans_query_index_oids"

typedef void *(*deserialize_hook_type)(void *, void *);
void *jsonb_to_node_tree(Jsonb *json, deserialize_hook_type hook_ptr, void *context);

Jsonb *node_tree_to_jsonb(const void *obj, Oid fake_func, bool skip_location_from_node);
void common_walker(const void *obj, void (*callback)(void *));

/*
 * MakeTupleTableSlot()
 */
#if PG_VERSION_NUM >= 110000
#define MakeTupleTableSlotCompat() \
	MakeTupleTableSlot(NULL)
#else
#define MakeTupleTableSlotCompat() \
	MakeTupleTableSlot()
#endif

#if PG_VERSION_NUM >= 100000
#define index_insert_compat(rel, v, n, t, h, u) \
	index_insert(rel, v, n, t, h, u, BuildIndexInfo(rel))
#else
#define index_insert_compat(rel, v, n, t, h, u) index_insert(rel, v, n, t, h, u)
#endif

#ifndef PG_GETARG_JSONB_P
#define PG_GETARG_JSONB_P(x) PG_GETARG_JSONB(x)
#endif

#ifndef PG_RETURN_JSONB_P
#define PG_RETURN_JSONB_P(x) PG_RETURN_JSONB(x)
#endif

#ifndef DatumGetJsonbP
#define DatumGetJsonbP(d) DatumGetJsonb(d)
#endif

enum
{
	Anum_sr_query_hash = 1,
	Anum_sr_query_id,
	Anum_sr_plan_hash,
	Anum_sr_enable,
	Anum_sr_query,
	Anum_sr_plan,
	Anum_sr_reloids,
	Anum_sr_index_reloids,
	Anum_sr_attcount
} sr_plans_attributes;

#endif
