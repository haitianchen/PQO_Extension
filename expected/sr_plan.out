CREATE EXTENSION sr_plan;
SET sr_plan.log_usage = NOTICE;
CREATE TABLE test_table(test_attr1 int, test_attr2 int);
INSERT INTO test_table SELECT i, i + 1 FROM generate_series(1, 20) i;
SET sr_plan.write_mode = true;
SELECT * FROM test_table WHERE test_attr1 = _p(10);
NOTICE:  sr_plan: collected parameter on 44
NOTICE:  sr_plan: saved plan for SELECT * FROM test_table WHERE test_attr1 = _p(10);
 test_attr1 | test_attr2 
------------+------------
         10 |         11
(1 row)

SELECT * FROM test_table WHERE test_attr1 = 10;
NOTICE:  sr_plan: saved plan for SELECT * FROM test_table WHERE test_attr1 = 10;
 test_attr1 | test_attr2 
------------+------------
         10 |         11
(1 row)

SELECT * FROM test_table WHERE test_attr1 = 10;
 test_attr1 | test_attr2 
------------+------------
         10 |         11
(1 row)

SET sr_plan.write_mode = false;
SELECT * FROM test_table WHERE test_attr1 = _p(10);
NOTICE:  sr_plan: collected parameter on 44
 test_attr1 | test_attr2 
------------+------------
         10 |         11
(1 row)

SELECT * FROM test_table WHERE test_attr1 = 10;
 test_attr1 | test_attr2 
------------+------------
         10 |         11
(1 row)

SELECT * FROM test_table WHERE test_attr1 = 15;
 test_attr1 | test_attr2 
------------+------------
         15 |         16
(1 row)

UPDATE sr_plans SET enable = true;
SELECT * FROM test_table WHERE test_attr1 = _p(10);
NOTICE:  sr_plan: collected parameter on 44
NOTICE:  sr_plan: restored parameter on 44
NOTICE:  sr_plan: cached plan was used for query: SELECT * FROM test_table WHERE test_attr1 = _p(10);
 test_attr1 | test_attr2 
------------+------------
         10 |         11
(1 row)

SELECT * FROM test_table WHERE test_attr1 = _p(15);
NOTICE:  sr_plan: collected parameter on 44
NOTICE:  sr_plan: restored parameter on 44
NOTICE:  sr_plan: cached plan was used for query: SELECT * FROM test_table WHERE test_attr1 = _p(15);
 test_attr1 | test_attr2 
------------+------------
         15 |         16
(1 row)

SELECT * FROM test_table WHERE test_attr1 = 10;
NOTICE:  sr_plan: cached plan was used for query: SELECT * FROM test_table WHERE test_attr1 = 10;
 test_attr1 | test_attr2 
------------+------------
         10 |         11
(1 row)

SELECT * FROM test_table WHERE test_attr1 = 15;
 test_attr1 | test_attr2 
------------+------------
         15 |         16
(1 row)

SELECT enable, query FROM sr_plans ORDER BY length(query);
 enable |                        query                        
--------+-----------------------------------------------------
 t      | SELECT * FROM test_table WHERE test_attr1 = 10;
 t      | SELECT * FROM test_table WHERE test_attr1 = _p(10);
(2 rows)

DROP TABLE test_table;
SELECT enable, query FROM sr_plans ORDER BY length(query);
 enable | query 
--------+-------
(0 rows)

CREATE TABLE test_table(test_attr1 int, test_attr2 int, test_attr3 int);
SELECT * FROM test_table WHERE test_attr1 = _p(10);
NOTICE:  sr_plan: collected parameter on 44
 test_attr1 | test_attr2 | test_attr3 
------------+------------+------------
(0 rows)

SELECT * FROM test_table WHERE test_attr1 = 10;
 test_attr1 | test_attr2 | test_attr3 
------------+------------+------------
(0 rows)

SELECT * FROM test_table WHERE test_attr1 = 10;
 test_attr1 | test_attr2 | test_attr3 
------------+------------+------------
(0 rows)

SELECT * FROM test_table WHERE test_attr1 = 15;
 test_attr1 | test_attr2 | test_attr3 
------------+------------+------------
(0 rows)

DROP EXTENSION sr_plan CASCADE;
NOTICE:  sr_plan was disabled
DROP TABLE test_table;
